src/otg/method/window_based_clumping.py

```diff
@@ -181,158 +184,183 @@ def _find_peak(position: ndarray, window_size: int) -> DenseVector:
+                # Adding study-locus id:
+                .withColumn(
+                    "studyLocusId",
+                    StudyLocus.assign_study_locus_id("studyId", "variantId"),
                 )
             ),
-            False,
+            _schema=StudyLocus.get_schema(),
         )
 
     @classmethod
-    def clump(
+    def clump_with_locus(
         cls: type[WindowBasedClumping],
         summary_stats: SummaryStatistics,
         window_length: int,
+        p_value_significance: float = 5e-8,
+        p_value_baseline: float = 0.05,
     ) -> StudyLocus:
-        """Clump summary statistics by distance.
+        """Clump significant associations while collecting locus around them.
 
         Args:
-            summary_stats (SummaryStatistics): summary statistics to clump
-            window_length (int): window length in basepair
+            cls (type[WindowBasedClumping]): _description_
+            summary_stats (SummaryStatistics): Input summary statistics dataset
+            window_length (int): Window size in  bp, used for distance based clumping and collecting locus
+            p_value_significance (float, optional): GWAS significance threshold used to filter peaks. Defaults to 5e-8.
+            p_value_baseline (float, optional): Least significant threshold. Below this, all snps are dropped. Defaults to 0.05.
 
         Returns:
-            StudyLocus: clumped summary statistics
+            StudyLocus: _description_
         """
-        return StudyLocus(
-            _df=summary_stats.df.withColumn(
-                "cluster_id",
-                # First identify clusters of variants within the window
-                WindowBasedClumping._identify_cluster_peaks(
-                    f.col("studyId"),
-                    f.col("chromosome"),
-                    f.col("position"),
-                    window_length,
-                ),
-            ).groupBy("cluster_id")
-            # Aggregating all data from each cluster:
-            .agg(
-                WindowBasedClumping._collect_clump(
-                    f.col("pValueMantissa"), f.col("pValueExponent")
-                ).alias("clump")
+        # Exclude problematic regions from clumping:
+        filtered_summary_stats = reduce(
+            lambda df, region: df.exclude_region(region),
+            cls.EXCLUDED_REGIONS,
+            summary_stats,
+        )
+
+        # Run distance based clumping on the summary stats:
+        clumped_dataframe = WindowBasedClumping.clump(
+            filtered_summary_stats,
+            window_length=window_length,
+            p_value_significance=p_value_significance,
+        ).df.alias("clumped")
+
+        # Extract column names:
+        columns = filtered_summary_stats.df.columns
+
+        # Dropping variants not meeting the baseline criteria:
+        sumstats_baseline = filtered_summary_stats.pvalue_filter(p_value_baseline).df
+
+        sumstats_baseline_renamed = (
```

Comment:
If I understand correctly, you are applying reduce to perform the renaming operation as many times as columns we have. 
As a suggestion, i think you can achieve the same with a single select statement. I'm more used to this solution, so for me it is more readable, but please go for whatever you like best

```suggestion
        sumstats_baseline_renamed = sumstats_baseline.selectExpr(*[f"{col} as tag_{col}" for col in sumstats_baseline.columns])

```

Reply:
This is great.